app/victoria-traces/servicegraph/servicegraph.go

```diff
@@ -0,0 +1,107 @@
+	"github.com/VictoriaMetrics/VictoriaTraces/app/vtstorage"
+)
+
+var (
+	enableServiceGraphTask     = flag.Bool("servicegraph.enableTask", false, "Whether to enable background task for generating service graph. It should only be enabled on VictoriaTraces single-node or vtstorage.")
+	serviceGraphTaskInterval   = flag.Duration("servicegraph.taskInterval", time.Minute, "The background task interval for generating service graph data. It requires setting -servicegraph.enableTask=true.")
+	serviceGraphTaskTimeout    = flag.Duration("servicegraph.taskTimeout", 30*time.Second, "The background task timeout duration for generating service graph data. It requires setting -servicegraph.enableTask=true.")
+	serviceGraphTaskLookbehind = flag.Duration("servicegraph.taskLookbehind", time.Minute, "The lookbehind window for each time service graph background task run. It requires setting -servicegraph.enableTask=true.")
+	serviceGraphTaskLimit      = flag.Uint64("servicegraph.taskLimit", 1000, "How many service graph relations each task could fetch for each tenant. It requires setting -servicegraph.enableTask=true.")
+)
+
+var (
+	sgt *serviceGraphTask
+)
+
+func Init() {
+	if *enableServiceGraphTask {
+		sgt = newServiceGraphTask()
+		sgt.Start()
+	}
+}
+
+func Stop() {
+	if *enableServiceGraphTask {
+		sgt.Stop()
+	}
+}
+
+type serviceGraphTask struct {
+	stopCh chan struct{}
+}
+
+func newServiceGraphTask() *serviceGraphTask {
+	return &serviceGraphTask{
+		stopCh: make(chan struct{}),
+	}
+}
+
+func (sgt *serviceGraphTask) Start() {
+	logger.Infof("starting background task for service graph, interval: %v, lookbehind: %v", *serviceGraphTaskInterval, *serviceGraphTaskLookbehind)
+	ticker := time.NewTicker(*serviceGraphTaskInterval)
+	go func() {
+		for {
+			select {
+			case <-sgt.stopCh:
+				return
+			case <-ticker.C:
+				ctx, cancelFunc := context.WithTimeout(context.Background(), *serviceGraphTaskTimeout)
+				GenerateServiceGraphTimeRange(ctx)
+				cancelFunc()
+			}
+		}
+	}()
+}
+
+func (sgt *serviceGraphTask) Stop() {
+	close(sgt.stopCh)
+}
+
+func GenerateServiceGraphTimeRange(ctx context.Context) {
+	endTime := time.Now().Truncate(*serviceGraphTaskInterval)
+	startTime := endTime.Add(-*serviceGraphTaskLookbehind)
+
+	tenantIDs, err := vtstorage.GetTenantIDsByTimeRange(ctx, startTime.UnixNano(), endTime.UnixNano())
+	if err != nil {
+		logger.Errorf("cannot get tenant ids: %s", err)
+		return
+	}
+
+	// query and persist operations are executed sequentially, which helps not to consume excessive resources.
+	for _, tenantID := range tenantIDs {
+		// Build a fake HTTP *Request. It helps align the way of handling tenant-related input
+		// in both vtselect and vtinsert, as they assume tenant info will exist in HTTP headers.
+		r, _ := http.NewRequestWithContext(ctx, "", "", nil)
+		r.Header.Set("AccountID", strconv.FormatUint(uint64(tenantID.AccountID), 10))
+		r.Header.Set("ProjectID", strconv.FormatUint(uint64(tenantID.ProjectID), 10))
```

Comment:
The dependency on the HTTP request in this background job seems out of place. Consider removing it and passing `CommonParams` struct instead.